home_audit.ts

```diff
@@ -0,0 +1,691 @@
+}
+
+function getBuildingConditionedFootprint(about, zone) {
+    const {zone_floor} = zone;
+    const {conditioned_floor_area, num_floor_above_grade} = about;
+    let conditioned_basement_area = 0;
+    // For conditioned footprint, we need to subtract the area of any conditioned basement floors
+    zone_floor.filter((floor) => (
+        floor.foundation_type === 'cond_basement'
+    )).forEach((floor) => (
+        conditioned_basement_area += floor.floor_area
+    ));
+    const footprint_area = conditioned_floor_area - conditioned_basement_area;
+    return Math.floor(footprint_area / num_floor_above_grade);
+}
+
+/**
+ * Get the Cross validation messages for the system of the JSON Schema
+ * @param {object} systems
+ */
+function getSystemCrossValidation(systems) {
+    const {hvac, domestic_hot_water, generation} = systems;
+    if(hvac) {
+        checkHvacFraction(hvac);
+        hvac.forEach((hvac_system, index) => {
+            checkHeatingCoolingTypeValid(hvac_system, index);
+            checkHeatingEfficiencyValid(hvac_system, index);
+            checkCoolingEfficiencyValid(hvac_system, index);
+            checkSystemYearValid(hvac_system, index);
+            checkHvacDistribution(hvac_system, index);
+        });
+    }
+    if(domestic_hot_water) {
+        checkHotWaterCategoryValid(domestic_hot_water, hvac);
+        checkHotWaterFuelValid(domestic_hot_water);
+        checkHotWaterEfficiencyValid(domestic_hot_water);
+        checkHotWaterYearValid(domestic_hot_water);
+        checkHotWaterEnergyFactorValid(domestic_hot_water);
+    }
+    if(generation && generation.solar_electric) {
+        checkSolarElectricYearValid(generation.solar_electric);
+    }
+}
+
+/**
+ * Check that the HVAC fraction is equal to 1 (100%)
+ */
+function checkHvacFraction(hvac) {
+    let total_fraction = 0;
+    hvac.forEach((hvac_system) => {
+        if(hvac_system.hvac_fraction) {
+            total_fraction += hvac_system.hvac_fraction;
+        }
+    });
+    if(total_fraction !== 1) {
+        hvac.forEach((hvac_system, index) => {
+            if (hvac_system.hvac_fraction) {
+                addErrorMessage(`systems/hvac/${index}/hvac_fraction`, `Total HVAC Fraction must equal 100%`);
+            }
+        });
+    }
+}
+
+/**
+ * Check that the heating and cooling methods are compatible
+ */
+const HEATING_FUEL_TO_TYPE = {
+    'natural_gas': [
+        'central_furnace',
+        'wall_furnace',
+        'boiler'
+    ],
+    'lpg': [
+        'central_furnace',
+        'wall_furnace',
+        'boiler'
+    ],
+    'fuel_oil': [
+        'central_furnace',
+        'wall_furnace',
+        'boiler'
+    ],
+    'electric': [
+        'central_furnace',
+        'heat_pump',
+        'mini_split',
+        'gchp',
+        'baseboard',
+        'boiler'
+    ],
+    'cord_wood': ['wood_stove'],
+    'pellet_wood': ['wood_stove'],
+}
```

Comment:
Is this logic not handled in the Schema?  It is totally capable of it, I think

Reply:
I don't think it is - I couldn't find anything that looked like it was doing this kind of mapping logic.